accepted/cfi-improvements-with-pauth-and-bti.md

```diff
@@ -0,0 +1,142 @@
+
+## CFI improvements to assembly, C, C++, and Rust code
+
+Improving CFI for compiled C, C++, and Rust code with the same technologies is outside the scope of
+this proposal, but in general it should be achievable by passing the appropriate parameters to the
+respective compiler.
+
+Functions implemented in assembly will get a similar treatment as generated code, i.e. they will
+start with the `PACIASP` instruction. However, the regular return will be preserved and instead will
+be preceded by the `AUTIASP` instruction. The reason is that both `AUTIASP` and `PACIASP` act as
+`NOP` instructions when executed by a processor that does not support PAuth, thus making the
+assembly code generic.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Since the existing implementation already uses the standard back-edge CFI techniques that are
+preferred in the absence of special hardware support (i.e. a separate protected stack that is not
+used for buffers that could be accessed out of bounds), the alternative is not to implement the
+proposal, so the rationale is based mainly on the overhead being insignificant. In terms of code
+size the impact of the back-edge CFI improvements is an additional instruction per function, or 2
+for functions implemented in assembly.
+
+The [Clang CFI design][clang-cfi-design] provides an idea for an alternative implementation of the
+forward-edge CFI mechanism that is enabled by BTI. It involves instrumenting every indirect branch
+to check if its destination is permitted. While the overhead of this approach can be reduced by
+using efficient data structures for the destination address lookup and optionally limiting the
+checks only to indirect function calls, it is still significantly larger than the worst-case BTI
+overhead of one instruction per basic block per function. On the other hand, it does not require any
+special hardware support, so it could be applied to all supported platforms.
+
+[clang-cfi-design]: https://clang.llvm.org/docs/ControlFlowIntegrityDesign.html
+
+# Open questions
+[open-questions]: #open-questions
+
+- What is the performance overhead of the proposal?
+- What technologies are available in other instruction set architectures to achieve the same goals?
+- What hardening approaches are applicable to the fiber implementation? The fiber switching code
+saves the values of all callee-saved registers on the stack, i.e. memory that is potentially
+accessible to an attacker. Some of those values could be code addresses that would be used by
+indirect branches, so should we devise a scheme to authenticate them? While the regular pointer
+authentication instructions assume that they are operating on valid virtual addresses (which implies
+that the most significant bits are redundant and could be repurposed), PAuth provides operations to
+authenticate arbitrary data, which could be used in this case.
```

Comment:
I investigated this issue a bit, and I have some further remarks:

> ... it sounds like we'll probably need to at least have some updates to the aarch64 implementation of `wasmtime-fiber`...

Actually we do not need to change anything to make things work (refer to bytecodealliance/wasmtime#3606), but that would leave a gap in the CFI implementation.

> ... but setjmp/longjmp would otherwise be handled by the system libraries?

Yes, that is correct, and it looks like on Linux the system libraries do not do anything except maybe authenticate the return address, but I do not think that it is by design - maybe it is just an oversight.

As for Apple's recommendation, it appears that I was partially right - it is written in the context of their pointer authentication ABI variant as far as I can tell, which mandates that all pointers be signed immediately after being generated, so no raw pointers would be kept in registers (or writable memory). In that situation it makes sense that only the set of core registers mentioned above needs special handling. We would not have the same guarantee in our stack switching code initially, so we should consider something more sophisticated, but we could start with the same approach, especially since it should be easier to implement - it could be done with just the `NOP` versions of the instructions, I think.